src/conversion/view-selection-to-model-converters.js

```diff
@@ -34,7 +35,12 @@ export function convertSelectionChange( modelDocument, mapper ) {
 		const ranges = [];
 
 		for ( const viewRange of viewSelection.getRanges() ) {
-			ranges.push( mapper.toModelRange( viewRange ) );
+			const mappedRange = mapper.toModelRange( viewRange );
+
+			const correctModelRangeStart = modelDocument.getNearestSelectionRange( mappedRange.start ).start;
+			const correctModelRangeEnd = modelDocument.getNearestSelectionRange( mappedRange.end ).end;
```

Comment:
I am surprised that we did not have this before. We should use view-to-model conversion to fix wrong selection. However, we need to ensure that `getNearestSelectionRange` do a good job here and do no break anything. Also, this code is not very readable and need docs.

Reply:
What do you mean by "we need to ensure that `getNearestSelectionRange` do a good job here"? This is not a new method and we already use it in couple of places. Seems to work fine :). The method - in theory - is very simple. It takes given position and looks for the nearest position (left and right) where a model selection can be placed. As I said, we already use this method and it solves our problems.

Reply:
I mean that now it will be called very often: whenever one type, move the selection, do... whatever. If this method will be called that ofter we should review it better than before ;) We need to ensure that it will not break any scenario. Also, we need to think about the efficiency.

Reply:
`getNearestSelectionRange()` may be used in a couple of places, but here we use it in a bit special and new context. And critical for things like stability with various features and performance.

Reply:
I understand your worries. Let's break it down.

Correctness. First of all, `getNearestSelectionRanges()` has quite a few tests written already. You can check them. I could have written a few tests for the conversion itself, but it would be mostly repeating those tests. In the PR I propose to use `getNearestSelectionRange()`  on a range instead of position. I've already wanted to introduce this to the method itself when solving another issue, but ultimately it was not needed. When you think about this solution it makes sense to take boundaries, expand them to ranges and then take appropriate boundaries. For example when selection is next to widget:
`foo<widget />[]` -> `foo{[<widget />]}` -> `foo{<widget />]` -> `foo[<widget />]`. This also works nicely for non-collapsed ranges. I haven't thought of all the cases, of course, but in theory it seems to be a good idea. I could write some test for non-collapsed ranges.

Efficiency. We have to do _any_ fixing. This is a fact. So we always have to check whether the selection is correct. I don't know of any correct and faster way than checking the schema. That's how we do it. As you can see, `getNearestSelectionRange()` has a check at the beginning which checks whether given position is correct in the first place. This is a minimal amount of processing that we have to do no matter what solution we choose. So we will do any extra processing only if the selection was in the wrong place. Which should not happen that often and should not have negative impact on editor smoothness.

Other worries. The only thing I am worried ATM is that we can get into some kind of stupid infinite loop. That deserves a second thought and a test. I am worried that this might happen if, for example there is no correct selection (editor is empty) or the browser tries to fix the selection in DOM all the time (but this is handled in another place).